File: runtime/src/de/fatox/meta/assets/MetaAssetProvider.java
package de.fatox.meta.assets;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.assets.AssetDescriptor;
import com.badlogic.gdx.assets.AssetManager;
import com.badlogic.gdx.assets.loaders.ModelLoader;
import com.badlogic.gdx.assets.loaders.TextureLoader;
import com.badlogic.gdx.files.FileHandle;
import com.badlogic.gdx.graphics.GL30;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.TextureAtlas;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.graphics.g3d.Model;
import com.badlogic.gdx.graphics.g3d.attributes.TextureAttribute;
import com.badlogic.gdx.scenes.scene2d.utils.Drawable;
import com.badlogic.gdx.scenes.scene2d.utils.TextureRegionDrawable;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.IntMap;
import com.badlogic.gdx.utils.ObjectMap;
import de.fatox.meta.api.AssetProvider;

public class MetaAssetProvider implements AssetProvider {
    private static TextureLoader.TextureParameter defaultTexParam = new TextureLoader.TextureParameter();
    private static ModelLoader.ModelParameters defaultModelParam = new ModelLoader.ModelParameters();
    private AssetManager assetManager = new AssetManager();
    private Array<TextureAtlas> atlasCache = new Array<>();
    private IntMap<Array<? extends TextureRegion>> animCache = new IntMap<>();

    private ObjectMap<String, XPKFileHandle> packFileCache = new ObjectMap<>();

    static {
        defaultTexParam.genMipMaps = true;
        defaultTexParam.minFilter = Texture.TextureFilter.MipMapLinearLinear;
        defaultModelParam.textureParameter = defaultTexParam;
    }

    @Override
    public boolean loadAssetsFromFolder(FileHandle folder) {
        if (folder.isDirectory()) {
            for (FileHandle itrHandle : folder.list()) {
                if (itrHandle.extension().equalsIgnoreCase(XPKLoader.EXTENSION)) {
                    Array<XPKFileHandle> list = XPKLoader.INSTANCE.getList(itrHandle);
                    list.forEach(it -> {
                        packFileCache.put(it.name(), it);
                    });
                }
            }
            return true;
        }
        return false;
    }

    @Override
    public <T> void load(String name, Class<T> type) {
        loadIntern(new AssetDescriptor(name, type));
    }


    private <T> void loadIntern(AssetDescriptor<T> descr) {
        if (descr.type == Model.class) {
            assetManager.load(descr.fileName, Model.class, defaultModelParam);
        } else if (descr.type == Texture.class && !descr.fileName.contains("ui")) {
            assetManager.load(descr.fileName, Texture.class, defaultTexParam);
        } else {
            assetManager.load(descr);
        }

        assetManager.finishLoading();
        if (descr.type == Model.class) {
            Model model = assetManager.get(descr.fileName, Model.class);
            TextureAttribute attribute = (TextureAttribute) model.materials.first().get(TextureAttribute.Diffuse);
            attribute.textureDescription.texture.bind();
            Gdx.gl30.glTexParameterf(GL30.GL_TEXTURE_2D, GL30.GL_TEXTURE_MAX_ANISOTROPY_EXT, 16);
        }

        if (descr.type == Texture.class) {
            Texture texture = assetManager.get(descr.fileName, Texture.class);
            texture.bind();
            Gdx.gl30.glTexParameterf(GL30.GL_TEXTURE_2D, GL30.GL_TEXTURE_MAX_ANISOTROPY_EXT, 16);
        }

        if (descr.type == TextureAtlas.class) {
            atlasCache.add(assetManager.get(descr.fileName, TextureAtlas.class));
        }
    }


    @Override
    public <T> T get(String name, Class<T> type, int index) {
        if (type == FileHandle.class) {
            if (packFileCache.containsKey(name)) {
                return (T) packFileCache.get(name);
            } else {
                return (T) Gdx.files.internal(name);
            }
        }
        if (assetManager.isLoaded(name, type)) {
            return assetManager.get(name, type);
        } else if (type == TextureRegion.class) {
            for (TextureAtlas atlas : atlasCache) {
                TextureAtlas.AtlasRegion region;
                if (index <= 0) {
                    region = atlas.findRegion(name);
                } else {
                    region = atlas.findRegion(name, index);
                }
                if (region != null) {
                    return (T) region;
                }
            }

            Texture texture = get(name, Texture.class);
            if (texture != null) {
                return (T) new TextureRegion(texture);
            }


        } else if (packFileCache.containsKey(name)) {
            XPKFileHandle xpkFileHandle = packFileCache.get(name);
            AssetDescriptor assetDescriptor = new AssetDescriptor(xpkFileHandle, type);
            loadIntern(assetDescriptor);
            return get(name, type);
        } else {
            load(name, type);
            return get(name, type);
        }
        return null;
    }

    public <T> T get(String name, Class<T> type) {
        return get(name, type, -1);
    }

    @Override
    public FileHandle get(String fileName) {
        return null;
    }


    @Override
    public Drawable getDrawable(String name) {
        return new TextureRegionDrawable(get(name, TextureRegion.class));
    }

    @Override
    public void finish() {
        assetManager.finishLoading();
    }

    /**
     * Returns a cached list of TextureRegions that represent the animation of the given texture
     *
     * @param baseName name of the texture
     * @param frames   limit frames of animations
     * @return
     */
    public Array<? extends TextureRegion> loadAnimationFrames(String baseName, int frames) {
        int key = baseName.hashCode();
        if (!animCache.containsKey(key)) {
            Array<TextureAtlas.AtlasRegion> regions = null;
            for (TextureAtlas atlas : atlasCache) {
                regions = atlas.findRegions(baseName);
                if (regions != null) {
                    continue;
                }
            }
            if (regions != null) {
                animCache.put(key, regions);
                regions.setSize(frames);
            }
        }
        return animCache.get(key);
    }

}

File: runtime/src/de/fatox/meta/assets/MetaAssetProvider.kt
package de.fatox.meta.assets

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.assets.AssetDescriptor
import com.badlogic.gdx.assets.AssetManager
import com.badlogic.gdx.assets.loaders.FileHandleResolver
import com.badlogic.gdx.assets.loaders.ModelLoader.ModelParameters
import com.badlogic.gdx.assets.loaders.TextureLoader.TextureParameter
import com.badlogic.gdx.files.FileHandle
import com.badlogic.gdx.graphics.GL30
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.TextureAtlas
import com.badlogic.gdx.graphics.g2d.TextureAtlas.AtlasRegion
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.graphics.g3d.Model
import com.badlogic.gdx.graphics.g3d.attributes.TextureAttribute
import com.badlogic.gdx.scenes.scene2d.utils.Drawable
import com.badlogic.gdx.scenes.scene2d.utils.TextureRegionDrawable
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.GdxRuntimeException
import com.badlogic.gdx.utils.IntMap
import com.badlogic.gdx.utils.ObjectMap
import de.fatox.meta.api.AssetProvider
import de.fatox.meta.api.ResourceKey
import de.fatox.meta.api.extensions.MetaLoggerFactory
import de.fatox.meta.api.extensions.debug
import de.fatox.meta.assets.XPKLoader.getList

private val log = MetaLoggerFactory.logger {}
private val defaultTexParam: TextureParameter = TextureParameter().apply {
	genMipMaps = true
	minFilter = Texture.TextureFilter.MipMapLinearLinear
}
private val defaultModelParam: ModelParameters = ModelParameters().apply { textureParameter = defaultTexParam }

class MetaAssetProvider : AssetProvider {
	private val assetManager = AssetManager(MetaFileHandleResolver())
	private val atlasCache = Array<TextureAtlas>()
	private val animCache = IntMap<Array<out TextureRegion>>()
	private val fileCache = ObjectMap<String, FileHandle>()

	override fun loadPackedAssetsFromFolder(folder: FileHandle): Boolean {
		if (folder.isDirectory) {
			for (itrHandle in folder.list()) {
				if (itrHandle.extension().equals(XPKLoader.EXTENSION, ignoreCase = true)) {
					val list = getList(itrHandle)
					list.forEach {
						fileCache.put(it.name(), it)
						fileCache.put(it.name().replace("/", "\\"), it)
						log.debug { "cache name: <${it.name()}>" }
					}
				}
			}
			return true
		}
		return false
	}

	override fun loadRawAssetsFromFolder(folder: FileHandle): Boolean {
		if (folder.isDirectory) {
			for (itrHandle in folder.list()) {
				if (itrHandle.isDirectory) {
					loadRawAssetsFromFolder(itrHandle)
				} else {
					fileCache.put(itrHandle.name(), itrHandle)
					fileCache.put(itrHandle.name().replace("/", "\\"), itrHandle)
				}
			}
			return true
		}
		return false
	}

	override fun <T: Any> load(name: String, type: Class<T>) {
		log.debug { "loading <$name>" }
		if (fileCache.containsKey(name)) {
			log.debug { "pack cache contains filename" }
			loadIntern(AssetDescriptor(fileCache[name], type))
		} else {
			loadIntern(AssetDescriptor(name, type))
		}
	}

	private fun <T: Any> loadIntern(descriptor: AssetDescriptor<T>) {
		when {
			descriptor.type == Model::class.java ->
				assetManager.load(descriptor.fileName, Model::class.java, defaultModelParam)
			descriptor.type == Texture::class.java && !descriptor.fileName.contains("ui") -> {
				log.debug { "ui load" }
				assetManager.load(descriptor.fileName, Texture::class.java, defaultTexParam)
			}
			else -> {
				log.debug { "normal load" }
				assetManager.load(descriptor)
			}
		}
		assetManager.finishLoading()
		if (descriptor.type == Model::class.java) {
			val model = assetManager.get(descriptor.fileName, Model::class.java)
			val attribute = model.materials.first()[TextureAttribute.Diffuse] as TextureAttribute
			attribute.textureDescription.texture.bind()
			Gdx.gl30.glTexParameterf(GL30.GL_TEXTURE_2D, GL30.GL_TEXTURE_MAX_ANISOTROPY_EXT, 16f)
		}
		if (descriptor.type == Texture::class.java) {
			val texture = assetManager.get(descriptor.fileName, Texture::class.java)
			texture.bind()
			Gdx.gl30.glTexParameterf(GL30.GL_TEXTURE_2D, GL30.GL_TEXTURE_MAX_ANISOTROPY_EXT, 16f)
		}
		if (descriptor.type == TextureAtlas::class.java) {
			atlasCache.add(assetManager.get(descriptor.fileName, TextureAtlas::class.java))
		}
	}

	override fun <T: Any> getResource(fileName: String, type: Class<T>, index: Int): T {
		return when {
			type == FileHandle::class.java -> {
				if (fileCache.containsKey(fileName))
					fileCache[fileName] as T
				else
					Gdx.files.internal(fileName) as T
			}
			assetManager.isLoaded(fileName, type) -> assetManager[fileName, type]
			type == TextureRegion::class.java -> {
				atlasCache.asSequence().map {
					if (index <= 0) it.findRegion(fileName) else it.findRegion(fileName, index)
				}.firstOrNull() as T? ?: getResource(fileName, Texture::class.java).let { TextureRegion(it) } as T?
			}
			fileCache.containsKey(fileName) -> {
				loadIntern(AssetDescriptor(fileCache[fileName], type))
				getResource(fileName, type)
			}
			else -> {
				load(fileName, type)
				getResource(fileName, type)
			}
		} ?: throw GdxRuntimeException("Resource not found: $fileName")
	}

	override fun getDrawable(name: String): Drawable {
		return TextureRegionDrawable(getResource(name, TextureRegion::class.java))
	}

	override fun finish() {
		assetManager.finishLoading()
	}

	override fun loadAnimationFrames(baseName: String, frames: Int): Array<out TextureRegion> {
		val key = 31 * baseName.hashCode() + frames
		if (!animCache.containsKey(key)) {
			// Since we use asSequence() map is lazily evaluated, thus only calling it when necessary.
			val regions: Array<AtlasRegion> = atlasCache.asSequence().map { it.findRegions(baseName) }.first()

			if (regions.size > 0) {
				if (frames > -1) regions.setSize(frames) // limit to the request number of frames
				animCache.put(key, regions)
			} else
				throw GdxRuntimeException("couldn't load $baseName")
		}
		return animCache[key]
	}

	internal inner class MetaFileHandleResolver : FileHandleResolver {
		override fun resolve(fileName: String): FileHandle {
			return fileCache[fileName] ?: Gdx.files.internal(fileName)
		}
	}
}